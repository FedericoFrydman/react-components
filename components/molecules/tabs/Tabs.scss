@import '../../assets/styles/colors.scss';
@import '../../assets/styles/shadows.scss';

.lsux-tab {
  > li {
    list-style: none;
  }
  &--bar,
  &--pills,
  &--tabs {
    > ul {
      margin-block-end: 0px;
      margin-block-start: 0px;
      padding-inline-start: 0px;
    }

    align-content: center;
    display: flex;
    width: fit-content;

    &.lsux-tab--stretch {
      width: auto;
      ul {
        display: flex;
        flex: 1;
        justify-content: center;
        text-align: center;
      }
    }

    a {
      color: $N700;
      text-decoration: none;

      &:hover {
        color: $B200;
        outline: none;
      }
    }
  }
  &--bar,
  &--tabs {
    a {
      color: $N700;
    }
  }
  &--pills {
    background-color: var(--brand-color-prim50);
    border: 1px solid var(--brand-color-prim100);
    border-radius: 50px;
    padding: 4px;

    &-selected {
      background-color: var(--brand-color-prim200);
      border-radius: 50px;
      color: $N00;
    }
    a {
      color: $N800;
    }
  }
  &--bar {
    border-bottom: 2px solid $N300;
    &-selected {
      border-bottom: 2px solid $N1000;
      color: $N1000;
    }
    .lsux-tab {
      background-clip: padding-box;
      margin-bottom: -2px;
      padding: 0;
    }
  }
  &--tabs {
    background-color: $N100;
    border-radius: 11px 11px 0px 0px;

    &-selected {
      background-color: $N00;
      border-radius: 11px 11px 0px 0px;
      color: $N1000;
      position: relative;
      &:before {
        background: var(--brand-color-prim200);
        border-radius: 10px 10px 0px 0px;
        content: '';
        flex: none;
        flex-grow: 0;
        height: 3px;
        left: 7%;
        order: 1;
        position: absolute;
        right: 0%;
        top: 0px;
        width: 85%;
        z-index: 1;
      }
    }

    h4 {
      font-weight: 400;
    }
  }
}
