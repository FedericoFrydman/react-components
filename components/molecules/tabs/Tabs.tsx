import { Ref, HTMLProps, forwardRef } from 'react';
import cn from 'classnames';

import './Tabs.scss';
import { Label } from '../../../components/atoms/label/Label';

export type TabsVariant = 'pills' | 'bar' | 'tabs';
export interface TabsProps extends Omit<HTMLProps<HTMLDivElement>, 'onClick' | 'css'> {
  /**
   * Names of the tabs
   */
  tabNames: string[];
  /**
   * Index in the array of the current active tab
   */
  activeIndex: number;
  /**
   * Define the styles of the tabs
   * @default pills
   */
  variant?: TabsVariant;
  /**
   * Stretch the component across the full width of its parent.
   * @default false
   */
  stretch?: boolean;
  /**
   * Custom classes for the component.
   */
  classNames?: cn.Argument;
  /**
   * Callback function when a tab is pressed. Parameter is the index of the tab clicked
   */
  onClick?: (index: number) => void;
  /**
   * The ref to the HTML DOM element.
   */
  ref?: Ref<HTMLDivElement>;
}

const TabsComponent = (
  { tabNames, activeIndex, onClick, variant = 'pills', stretch = false, classNames = [], ...props }: TabsProps,
  ref: TabsProps['ref'],
) => {
  if (variant === 'bar')
    console.warn('The bar variant in the Tabs component has been deprecated. Please use pills or tabs instead.');
  const classTab = 'lsux-tab';
  const cnTab = cn({ [`${classTab}--${variant}`]: variant, [`${classTab}--stretch`]: stretch }, classNames);

  return (
    <div {...props} className={cnTab} ref={ref}>
      {tabNames.map(function (name, index) {
        const handleClick = (event: React.MouseEvent<HTMLUListElement>): void => {
          event.preventDefault();
          onClick(index);
        };
        return (
          <ul
            key={index.toString()}
            className={cn({ [`${classTab}`]: true, [`${classTab}--${variant}-selected`]: index === activeIndex })}
            onClick={index === activeIndex ? null : handleClick}
          >
            <li className={cn('my-3', 'mx-4')}>
              {index === activeIndex ? (
                <Label text={name} labelSize="medium" />
              ) : (
                <a href="#">
                  <Label text={name} labelSize="medium" />
                </a>
              )}
            </li>
          </ul>
        );
      })}
    </div>
  );
};

/**
 * Tabs allow for an easy way to organize similar content into groups, without having to use separate pages to do so.
 * Tabs should be placed in a consistent location across the screens and render content without any page refresh.
 */

export const Tabs = forwardRef<HTMLDivElement, TabsProps>(TabsComponent);
