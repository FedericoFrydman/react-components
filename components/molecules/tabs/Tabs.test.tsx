import { createRef } from 'react';
import { render, screen } from '@testing-library/react';
import { shallow, ShallowWrapper } from 'enzyme';
import { Tabs, TabsProps, TabsVariant } from './Tabs';

describe('Tabs Component', () => {
  let testId: string;
  let wrapper: ShallowWrapper;

  beforeEach(() => {
    testId = 'test-tabs';
    wrapper = shallow(<Tabs activeIndex={2} tabNames={['One', 'Two', 'Three']} variant={'bar'} />);
  });

  test('should render the component', () => {
    expect(wrapper).toMatchSnapshot();
  });

  describe('ref', () => {
    const ref = createRef<HTMLDivElement>();

    it('provides a ref to the tabs', () => {
      expect(renderTabs({ ref })).toEqual(ref.current);
    });
  });

  describe('variant', () => {
    const variants: TabsVariant[] = ['pills', 'bar', 'tabs'];

    it(`defaults to 'pills'`, () => {
      const tabs = renderTabs();

      expect(tabs).toHaveClass(`lsux-tab--pills`, { exact: false });
    });

    function itAssignsTheRightClassForVariant(variant: TabsVariant) {
      const tabsTypeClass = `lsux-tab--${variant}`;

      describe(`when the variant is ${variant}`, () => {
        it(`assigns the ${tabsTypeClass} to the tabs`, () => {
          const tabs = renderTabs({ variant });

          expect(tabs).toHaveClass(tabsTypeClass);
        });
      });
    }

    variants.forEach((variant) => itAssignsTheRightClassForVariant(variant));
  });

  describe('stretch', () => {
    const stretchClass = `lsux-tab--stretch`;
    let stretch: boolean;

    describe('when stretch is true', () => {
      beforeEach(() => (stretch = true));

      it(`has the ${stretchClass} class`, () => {
        expect(renderTabs({ stretch })).toHaveClass(stretchClass, { exact: false });
      });
    });

    describe('when stretch is false', () => {
      beforeEach(() => (stretch = false));

      it(`does not have the ${stretchClass} class`, () => {
        expect(renderTabs({ stretch })).not.toHaveClass(stretchClass, { exact: false });
      });
    });
  });

  const renderTabs = (props: Omit<TabsProps, 'tabNames' | 'activeIndex'> = {}): HTMLElement => {
    render(<Tabs {...props} activeIndex={2} tabNames={['One', 'Two', 'Three']} data-testid={testId} />);
    return screen.getByTestId(testId);
  };
});
