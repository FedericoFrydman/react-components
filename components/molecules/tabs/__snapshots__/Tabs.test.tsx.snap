// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tabs Component should render the component 1`] = `
<div
  className="lsux-tab--bar"
>
  <ul
    className="lsux-tab"
    key="0"
    onClick={[Function]}
  >
    <li
      className="my-3 mx-4"
    >
      <a
        href="#"
      >
        <ForwardRef(LabelComponent)
          labelSize="medium"
          text="One"
        />
      </a>
    </li>
  </ul>
  <ul
    className="lsux-tab"
    key="1"
    onClick={[Function]}
  >
    <li
      className="my-3 mx-4"
    >
      <a
        href="#"
      >
        <ForwardRef(LabelComponent)
          labelSize="medium"
          text="Two"
        />
      </a>
    </li>
  </ul>
  <ul
    className="lsux-tab lsux-tab--bar-selected"
    key="2"
    onClick={null}
  >
    <li
      className="my-3 mx-4"
    >
      <ForwardRef(LabelComponent)
        labelSize="medium"
        text="Three"
      />
    </li>
  </ul>
</div>
`;
